File: years/2022/d04.py
import os.path

testing = False
day = '04'
if testing:
    f_input = os.path.join('inputs', f'input{day}-test.txt')
else:
    f_input = os.path.join('inputs', f'input{day}.txt')

def part1(input):
    with open(input) as f:
        data = [line.split(',') for line in f]
    counter = 0
    for elf1, elf2 in data:
        e1a, e1b = map(int, elf1.split('-'))
        e2a, e2b = map(int, elf2.split('-'))
        if e1a >= e2a and e1b <= e2b:
            counter += 1
        elif e2a >= e1a and e2b <= e1b:
            counter += 1
    return counter

print('Part 1:', part1(f_input))

def part2(input):
    with open(input) as f:
        data = [line.split(',') for line in f]
    counter = 0
    for elf1, elf2 in data:
        e1a, e1b = map(int, elf1.split('-'))
        e2a, e2b = map(int, elf2.split('-'))
        if e2a <= e1a <= e2b or e2a <= e1b <= e2b:
            counter += 1
        elif e1a <= e2a <= e1b or e1a <= e2b <= e1b:
            counter += 1
    return counter

print('Part 2:', part2(f_input))

File: years/2022/d04.rb
# frozen_string_literal: true

testing = false
day = '04'

if testing
  f_input = File.join('inputs', "input#{day}-test.txt")
else
  f_input = File.join('inputs', "input#{day}.txt")
end

def part1(input)
  pairs = []
  File.foreach(input) { |pair| pairs << pair.chomp.split(',') }
  count = 0

  pairs.each do |elf1, elf2|
    e1a, e1b = elf1.split('-').map(&:to_i)
    e1range = e1a..e1b

    e2a, e2b = elf2.split('-').map(&:to_i)
    e2range = e2a..e2b

    count += 1 if e1range.cover?(e2range) || e2range.cover?(e1range)
  end

  count
end

puts("Part 1: #{part1(f_input)}")

def part2(input)
  pairs = []
  File.foreach(input) { |pair| pairs << pair.chomp.split(',') }
  count = 0

  pairs.each do |elf1, elf2|
    e1a, e1b = elf1.split('-').map(&:to_i)
    e1range = e1a..e1b

    e2a, e2b = elf2.split('-').map(&:to_i)
    e2range = e2a..e2b

    count += 1 if e1range.include?(e2a) || e1range.include?(e2b) || e2range.include?(e1a) || e2range.include?(e1b)
  end

  count
end

puts ("Part 2: #{part2(f_input)}")
